thunder/benchmarks/benchmark_litgpt.py

```diff
@@ -307,13 +285,47 @@ def setup_distributed(self, model):
                     sharding_strategy=sharding_strategy,
                     bucketing_strategy=bucketing_strategy,
                 )
+            else:
+                if self.distributed_mode == "fsdp2":
+                    raise ValueError(f"`fsdp2` is not available for `thunder.jit`.")
         else:
             if self.distributed_mode == "ddp":
                 model = torch.nn.parallel.DistributedDataParallel(
                     model,
                     device_ids=[local_rank],
                     bucket_cap_mb=self.ddp_bucket_size,
                 )
+            elif self.distributed_mode == "fsdp2":
+                # reference: https://github.com/pytorch/torchtitan/blob/6e7a183/docs/fsdp.md
+                from torch.distributed._composable.fsdp import fully_shard, MixedPrecisionPolicy
+
+                torch.cuda.set_device(local_rank)
+                mesh = None
+                if self.sharding_size is not None:
+                    mesh = init_device_mesh("cuda", (int(world_size / self.sharding_size), self.sharding_size))
+                else:
+                    mesh = init_device_mesh("cuda", (world_size,))
+
+                reshard_after_forward: bool = self.shard_mode == "zero3"
+
+                for transformer_block in model.modules():
+                    if isinstance(transformer_block, Block):
+                        fully_shard(
+                            transformer_block,
+                            mesh=mesh,
+                            reshard_after_forward=reshard_after_forward,
+                            mp_policy=MixedPrecisionPolicy(),
+                        )
+
+                fully_shard(
+                    model,
+                    mesh=mesh,
+                    reshard_after_forward=reshard_after_forward,
+                    mp_policy=MixedPrecisionPolicy(),
+                )
+                model.to_empty(device=self.device)
+                model.apply(model._init_weights)
```

Comment:
Before these two lines the parameters are on the 'meta' device, right?

Reply:
ys